clippy_lints/src/needless_move.rs

```diff
@@ -0,0 +1,222 @@
+//! This lint works by looking at the `min_captures` that `rustc` uses,
+//! and checks that for the expression `capture_kind_expr_id`, it would
+//! actually borrow normally, if it weren't for the move keyword.
+//!
+//! In such cases, the move keyword changes the semantics of the code (e.g.
+//! without it that capture would be a normal by reference capture, but with
+//! move it would get captured by value, and therefore we do not remove the `move`
+//! keyword from the closure).
+//!
+//! A small caveat for the approach above:
+//! There's both a borrow and a move of the same value into the closure, e.g.:
+//!
+//! ```no_run
+//! let x = String::new();
+//! let closure = move || {
+//!     let s = x.as_str(); // L1
+//!     println!("{s}");
+//!     drop(x); // L2
+//! };
+//! ```
+//!
+//! In this case, the `x` `String` gets moved into the closure (because of L2), but
+//! it is also borrowed prior to that at L1.
+//!
+//! `rustc`, in the presence of the `move` keyword automatically assumes that if
+//! it borrows a value, it's going to move it into the closure (in the example above at L1,
+//! so `capture_kind_expr_id` would point to the use on L1), but here, in the case
+//! of this lint, we should behave a little differently, namely we should first look
+//! at all the locations where a place is captured, and if any of them actually moves it,
+//! the closure would consume the value.
+//!
+//! The logic for this is handled in `MovedVariablesCtxt::get_required_kind`, where we
+//! try to infer the actual min capture kind needed.
+
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::sugg::DiagnosticExt;
+use rustc_errors::Applicability;
+use rustc_hir::{CaptureBy, Closure, Expr, ExprKind, HirId};
+use rustc_hir_typeck::expr_use_visitor as euv;
+use rustc_infer::infer::TyCtxtInferExt;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::mir::FakeReadCause;
+use rustc_middle::ty;
+use rustc_middle::ty::UpvarCapture;
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for closures and `async` blocks where the `move` is not necessary.
+    /// E.g. all the values are captured by value into the closure / `async` block.
+    ///
+    /// ### Why is this bad?
+    /// Pedantry
+    /// ### Example
+    /// ```no_run
+    /// let a = String::new();
+    /// let closure = move || {
+    ///     drop(a);
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let a = String::new();
+    /// let closure = || {
+    ///     drop(a);
+    /// };
+    /// ```
+    #[clippy::version = "1.76.0"]
+    pub NEEDLESS_MOVE,
+    restriction,
```

Comment:
Is there any disadvantage to applying this lint? If not, could this be `pedantic` instead?